Project templates for projects

oliverguenther opened the pull request:
- [x] Add templated property to project
- [x] Add templated filter to project query
- [x] Add button in toolbar to turn project into template
- [x] Make template selectable for users with `:copy_projects` permission
- [x] Call separate service that validates the request and schedules a copy job


**Open**
- Do we want to transport the status to the frontend similar to how exports do it? Currently, this PR simply uses a copy job that will send an email once ready

oliverguenther committed:
Add templated property to projects

oliverguenther committed:
Add templated filter to project query

fixup! Add templated filter to project query

oliverguenther committed:
Pass contract_options to SetAttributesService

This is needed for the contract to pick up an extra attribute passed through its options

oliverguenther committed:
Allow to select template in new form

oliverguenther committed:
Create separate service/contract for instantiation of templates

oliverguenther committed:
Move create specs from legacy to controller

oliverguenther committed:
Add project template factory

oliverguenther committed:
Add different handling to project creation if using template

oliverguenther committed:
Run CopyProjectJob for instantiating template with pararms

oliverguenther committed:
Pass contract_options in stubs

oliverguenther committed:
Fix create project spec

oliverguenther committed:
Add specs for instantiation

oliverguenther committed:
Fix icon size in styleguide

oliverguenther committed:
Toggle template button in settings

oliverguenther committed:
Add spec for toggle button

ulferts commented on app/contracts/projects/base_contract.rb:
```diff
@@ -36,6 +36,7 @@ class BaseContract < ::ModelContract
     attribute :name
     attribute :identifier
     attribute :description
+    attribute :templated
```
Changes to this attribute should be restricted to admins only so we should enforce this in the contract.

ulferts commented on app/helpers/projects_helper.rb:
```diff
@@ -145,6 +146,13 @@ def project_options_for_status(project)
     end
   end
 
+  def project_options_for_templated
```
This belongs into the contract. The contract should then double check whether the provided template_id is included in this set. This might not be strictly necessary for now but it is the general pattern we should follow (available options in the contract -> contract checks whether current value is included in the available options) and will allow us to easily and faultlessly adopt new features, e.g. specific permission to instantiate a template.

ulferts commented on app/views/projects/form/_select_template.html.erb:
```diff
@@ -0,0 +1,38 @@
+++#%>
+
+<div class="form--field">
+  <%= styled_label_tag 'from_template', l('project.template.use_template') %>
```
As this is a new template, prefer to use `t` over `l` so that we will be able to get rid of `l` eventually.

ulferts reviewed (changes requested):
I couldn't test all of the changes as I ran into errors rather quickly. They might be related to something on my machine as I see the tests being green. Let me know if you want to debug together.

We might want to add some description to the field at least until the first template is defined (and we think the user got the gist). 

![image](https://user-images.githubusercontent.com/617519/83243138-c9744880-a19d-11ea-8f53-66d8264a20ea.png)

~~I haven't been able to create a new project (not from template)~~ (my bad, didn't pull correctly):
![image](https://user-images.githubusercontent.com/617519/83243751-a0a08300-a19e-11ea-96de-4ceae7b1ebdb.png)
(my bad, didn't pull correctly

~~When attempting to create a project from template, I ran into a NoMethodError~~(same as above:
![image](https://user-images.githubusercontent.com/617519/83244019-0f7ddc00-a19f-11ea-89e2-2b25ce046843.png)

The project templates are presented in a flat list:
![image](https://user-images.githubusercontent.com/617519/83262753-c9844080-a1bd-11ea-8096-28978a41c363.png)
We already have a drop down listing projects in hierarchy (used for parent select). I think we should reuse that.

The "Template project" label is not linked to the checkbox so clicking on it does not work
![image](https://user-images.githubusercontent.com/617519/83262890-ff292980-a1bd-11ea-89b8-02be9a2e8e96.png)


> Do we want to transport the status to the frontend similar to how exports do it? Currently, this PR simply uses a copy job that will send an email once ready

I'd love this but think we should add a generic API endpoint to fetch the status so that we do not have to repeat ourselves. I did not do that when writing the status feature due to time constraints but I firmly believe this is the way to go.

Todos 
* [ ] Address comments in code
* [ ] Include the property in the projects api incl. documentation
* [ ] Spec the added template attribute in the project contract specs
* [ ] Reuse hierarchical project dropdown for template selection
* [ ] Link label to field

Open
* Do we need to seed a template?

ulferts commented:
I updated my initial review. Functionality looks mostly fine now, just some minor things :+1:

oliverguenther committed:
Ensure templated attribute is changeable only by admins

oliverguenther committed:
Use t instead of l helper

oliverguenther committed:
Move available templates method into contract

oliverguenther committed:
Ensure we set templated to false when copying

oliverguenther commented:
I believe to have addressed all remarks @ulferts . I also noticed that the copy was still marked as templated (which makes sense since it just takes all source attributes apart from those you have overwritten), so I made sure to set this to false in the calling service.

oliverguenther committed:
Better waiting for meeting title spec fix

ulferts reviewed (approved):
Looks fine. We only need some icons for the template toggle, now.
